python/ray/tune/search/vizier/vizier_search.py

```diff
@@ -0,0 +1,248 @@
+        metric: Optional[str] = None,
+        mode: Optional[str] = None,
+        algorithm: Optional[str] = 'DEFAULT',
+    ):
+        assert IMPORT_SUCCESSFUL, 'Vizier must be installed with `pip install google-vizier[jax]`.'
+        super(VizierSearch, self).__init__(metric=metric, mode=mode)
+        self._algorithm = algorithm
+
+        # For Vizier to identify the unique study.
+        self._study_id = f'ray_vizier_{uuid.uuid1()}'
+
+        # Mapping from Ray trial id to Vizier Trial client.
+        self._active_trials: Dict[str, clients.Trial] = {}
+
+        # Vizier service client.
+        self._study_client: Optional[clients.Study] = None
+
+        if isinstance(space, dict) and space:
+            resolved_vars, domain_vars, grid_vars = parse_spec_vars(space)
+            if resolved_vars:
+                raise TypeError(SPACE_ERROR_MESSAGE)
+            if domain_vars or grid_vars:
+                logger.warning(
+                    UNRESOLVED_SEARCH_SPACE.format(par="space", cls=type(self))
+                )
+                space = self.convert_search_space(space)
+            self._space = space
+        elif isinstance(space, svz.SearchSpace):
+            self._space = space
+        else:
+            raise TypeError(SPACE_ERROR_MESSAGE + " Got {}.".format(type(space)))
+
+    if self._space:
+        self._setup_vizier()
+    
+    def set_search_properties(
+        self, metric: Optional[str], mode: Optional[str], config: Dict, **spec
+    ) -> bool:
+        if self._study_client:  # The study is already configured.
+            return False
+
+        self._metric = metric or DEFAULT_METRIC
+        self._mode = mode
+        self._space = vzr.SearchSpaceConverter.to_vizier(config)
+   
+        self._setup_vizier()
+        return True
+
+    def _setup_vizier(self) -> None:
+        if self._mode == 'max':
+            vizier_goal = svz.ObjectiveMetricGoal.MAXIMIZE
+        elif self._mode == 'min:
+            vizier_goal = svz.ObjectiveMetricGoal.MINIMIZE
+        else:
+            raise RuntimeError(UNDEFINED_METRIC_MODE.format(cls=self.__class__.__name__, metric=self._metric, mode = self._mode))
+
+        study_config = svz.StudyConfig(
+            search_space=self._space,
+            algorithm=self._algorithm,
+            metric_information=[svz.MetricInformation(self._metric, goal=vizier_goal)],
+        )
+        self._study_client = clients.Study.from_study_config(
+            study_config, owner='raytune', study_id=self._study_id
+        )
+
+    def on_trial_result(self, trial_id: str, result: Dict) -> None:
+        if trial_id not in self._active_trials:
+            raise RuntimeError(f'No active trial for {trial_id}')
+        if self._study_client is None:
+            raise RuntimeError(
+                'VizierSearch not initialized! Set a search space first.'
+            )
+        trial_client = self._active_trials[trial_id]
```

Comment:
nit: Is this object a "trial" or a "client"? Looking over this I think this is a `Trial` object suggested by the `Study` client, but want to make sure I'm not missing something.
```suggestion
        client_trial = self._active_trials[trial_id]
```

Reply:
"Both".

We get this question a lot, and it's understandable - basically the regular user only interacts with Vizier objects using the client library: https://github.com/google/vizier/blob/main/vizier/client/client_abc.py

These are technically RPC client wrappers (so that the user doesn't need to deal with RPC logic themselves), but the corresponding actual material object (e.g. `vz.Trial`) can be obtained by e.g. `trial_client.materialize()`.

Agreed that the naming is a bit weird (e.g. `active_trials` would therefore technically be `active_trial_clients`) but we've gone too deep into this style of doing things.

So when I wrote it as `trial_client`, it means "the client which corresponds to that specific trial"

Does this seem OK to you for now?

Reply:
Oh yes this makes a lot of sense now, thanks for explaining!